### ISSUE TITLE
How can I trigger a Default value when inserting an unassigned std::optional?

### ISSUE THREAD

**Issue by Laro88** [jtv/libpqxx#409]
I haven't found a way to do this (yet) but I am hoping that it is possible. I can keep the two main flows in the code (checking IsValid() == true) and then have an extra parameter, however I was hoping that an unassigned std::optional would cause trigger the Default value.
```
#include <iostream>
#include <string>
#include <pqxx/pqxx>

int main()
{
	const std::string connstr = "user=postgres dbname=octopus password=postgres";

	try
	{
		{
			pqxx::connection c(connstr);
			pqxx::work w(c);
			w.exec("DROP TABLE iF EXISTS testoptionaluuid");
			w.exec("CREATE TABLE testoptionaluuid ( "
			       "idx serial primary KEY, "
			       "strx character varying(5) not NULL, "
			       "uuidx uuid DEFAULT uuid_generate_v1() unique not null " //I would like a std::optional<std::string> unassigned to trigger the insert
			       ");");

			w.commit();
			std::cout << "recreated table" << std::endl;
		}

		{
			pqxx::connection c(connstr);
			pqxx::work w(c);

			//check conn - just to be sure we are connected
			auto rNow = w.exec("SELECT now() as now");
			std::cout << "now " << rNow[0]["now"].as<std::string>() << std::endl;

			//works fine - we get a new uuid
			c.prepare("prepStatement", "INSERT INTO testoptionaluuid(strx) VALUES ($1) RETURNING idx, uuidx");
			std::string strX = "a";
			pqxx::result r = w.exec_prepared("prepStatement", strX);
			w.commit();
			std::cout << "executed " << r[0]["idx"] << " " << r[0]["uuidx"] << std::endl;
		}

		{
			//I would REALLY like a std::optional<std::string> optUuid to trigger the default creation as well
			pqxx::connection c(connstr);
			pqxx::work w(c);
			c.prepare("prepStatement", "INSERT INTO testoptionaluuid(strx, uuidx) VALUES ($1,$2) RETURNING idx, uuidx");
			std::string strX = "a";
			std::optional<std::string> optUuid;
			pqxx::result r = w.exec_prepared("prepStatement", strX, optUuid);
			w.commit();
			std::cout << "executed " << r[0]["idx"] << " " << r[0]["uuidx"] << std::endl;
		}
	}
	catch(const std::exception& e)
	{
		std::cout << e.what();
	}

	int j = 0;
}
```

**Comment by jtv** [comment]
There's no way yet, but you can create your own mechanism:

Create a C++ type for a default value.  Write a `string_traits` (see doc on
data types) such that converting a value of this type to a string simply
returns "DEFAULT".

I'll build this feature in when I can.  Dealing with some problems right
now, can't spend much time at the computer.

**Comment by tt4g** [comment]
If you use a function like `uuid_generate_v1`, I think `COALESCE` can solve the problem.

Example (Fiddle: http://sqlfiddle.com/#!17/ee7e5/1/0):

```sql
CREATE TABLE foo (
  foo UUID NOT NULL DEFAULT uuid_generate_v1()
);

INSERT INTO foo (foo) 
VALUES (COALESCE(NULL, uuid_generate_v1()));
```

This solution does not refer to `DEFAULT`, so you will need to update your SQL to match the schema change.

Another way: https://stackoverflow.com/questions/39165583/inserting-a-coalescenull-default

**Comment by jtv** [comment]
@Laro88 I just tried, and you can't pass `DEFAULT` to a prepared or parameterised statement.  And for a regular statement, there's just no point.  So I'd forget what I said about specialising `string_traits`, and just go with `COALESCE()` like @tt4g suggested.

**Comment by jtv** [comment]
@Laro88 is it working?

**Comment by Laro88** [comment]
I am actually in the middle of testing @jtv  to see if COALESCE works properly - with have issues with both datetime and uuid apparently so I have to investigate a bit.

**Comment by jtv** [comment]
Yes, `std::optional` is the preferred way.  But if your situation requires `std::variant`, for instance, you can use `nullptr_t` or `nullopt_t` as one of the possible contained types.  That way you don't have to wrap an `optional` inside your `variant`.

**Comment by KayEss** [comment]
Do you support [`std::monostate`](https://en.cppreference.com/w/cpp/utility/variant/monostate) as well for this? That's the way that `null` values are generally represented in variants.

**Comment by jtv** [comment]
Wasn't aware of `monostate` but shouldn't be hard.

**Comment by jtv** [comment]
Indeed it wasn't.  We support `std::monostate` as a null value now.

**Comment by Laro88** [comment]
Would it make sense to specialize string_traits for <std::optional<myownUuid>>

**Comment by jtv** [comment]
@Laro88 no, but if `myownUuid` is a class, then you'll probably want to specialise `string_traits<myownUuid>`.

There is a generic specialisation (that sounds like a contradiction in terms, but that's what it is) for `std::optional<T>` given any `T`.

**Comment by Laro88** [comment]
I will give the pqxx::string_traits<myownUuid> class a shot then @jtv 
std::optional<int> etc. works fine in queries with respect to default values, i was wondering if the std::optional<myownUuid> was the root cause of the problems experienced.

**Comment by Laro88** [comment]
i am a bit puzzled with the pqxx::string_traits in the into_buf function.
```
static char* into_buf(char* begin, char* end, systemobjects::DateTime const& value)
		{
			std::string s = "to_timestamp(" + std::to_string(value.InSeconds()) + ")";

			int len = s.size();

			//char* p = s.c_str();
			for(int j = 0; j < len ; ++j)
			{
				begin[j] = s.c_str()[j];
			}
			begin[len] = 0;
			return begin + len + 1;
		}
```
I have checked the memory, and the trailing 0 is present however I keep getting an exception:
```_Data = {_What=0x077590a8 "ERROR:  invalid input syntax for type timestamp: \"to_timestamp(1540008000)\"\n" ...}```

Postgres is complaining about the parameter provided
```
postgis_1         | 2021-02-05 11:00:52.236 UTC [4713] ERROR:  invalid input syntax for type timestamp: "to_timestamp(1540008000)"
postgis_1         | 2021-02-05 11:00:52.236 UTC [4713] STATEMENT:  INSERT INTO test_systemobjects_datetime(ts) VALUES($1) RETURNING id
```

I was hoping to be able to hide the internals of the DateTime object, this object hides some "interesting" mapping to ancient hardware - is this inline usage of to_timestamp being captured by the prepared statement in some way?

This little set of queries runs fine - just to test if the syntax was bugging out.
```
drop table if exists test_systemobjects_datetime;
create table test_systemobjects_datetime(id SERIAL, ts timestamp WITHOUT TIME ZONE);
prepare q(timestamp without time zone) as INSERT INTO test_systemobjects_datetime(ts) values($1);
execute q (to_timestamp(1540008000));
select * from test_systemobjects_datetime;
```

**Comment by tt4g** [comment]
"to_timestamp(1540008000)" appears to be interpreted as a single prepared parameter.
If it is interpreted as `to_timestamp($1)`, then it should be executed (PostgreSQL does not distinguish between functions and values contained in a single prepared parameter).

I think you can do it by changing the prepared statement to `INSERT INTO test_systemobjects_datetime(ts) values(to_timestamp($1))`.

And

```cpp
static char* into_buf(char* begin, char* end, systemobjects::DateTime const& value)
		{
			std::string s = value.InSeconds();

			int len = s.size();

			//char* p = s.c_str();
			for(int j = 0; j < len ; ++j)
			{
				begin[j] = s.c_str()[j];
			}
			begin[len] = 0;
			return begin + len + 1;
		}
```

**Comment by Laro88** [comment]
@tt4g this can be done  - but then the nice hiding work inside the string_traits implementation is lost.
It seems like libpqxx is guarding the query, since native direct queries runs fine.
```
prepare q(timestamp without time zone) as INSERT INTO test_mcmobjects_datetime(ts) values($1);
execute q (to_timestamp(1540008000));
```

**Comment by tt4g** [comment]
As far as I know, this is a limitation of libpq.
A string passed for `$1` in a prepared statement is not interpreted as a function.
A similar case I encountered was the LIKE clause.

**Comment by Laro88** [comment]
Thank you for the explanation @tt4g

**Comment by jtv** [comment]
@Laro88 if you're passing your value to a prepared or parameterised statement, then that statement will interpret the parameter as a string, not as SQL.  (And for good reason — you wouldn't want a user of a web app for instance to be able to type some SQL and have it executed on your database.)  So yes, as @tt4g says, you'll need to put the SQL code in the SQL statement, not in the value.

Some notes about the code:
 1. Don't use `std::to_string` for this — it will produce different formats depending on the user's locale.  But `pqxx::to_string` was designed not to have this problem.  (Sorry for the name confusion; in my defence, libpqxx defined a `to_string` before the standard did. ☺)
 2. Use 'single quotes' for SQL values, not "double quotes."
 3. For a useful little helper, have a look at `std::string_view::copy()`!

**Comment by jtv** [comment]
Oh, I'm sorry about the single-vs-double-quotes thing, I misread your code.  I'm recovering from a somewhat painful physical problem, and it's still very distracting.

**Comment by Laro88** [comment]
No problem @jtv  - hope your are recovering nicely.

I will use the pqxx:to_string instead, I saw your locale related warnings in the documentation.

The std::optional of my datetime object wrapper seems to work fine - just a couple of unittests more and then I will write a simplified git gist for others to start out from (your comments on it will be much appreciated).

I ended up using boost to get a decent string parameter in & out of string_traits, then the to_timestamp is gone from the insert queries, and the selects become nice as well.
